### ISSUE TITLE
SBT doesn't include system out or system error in JUnit XML reports

### ISSUE THREAD

**Issue by dmitchellmim** [sbt/sbt#6537]
## steps

1. Run JUnit tests with `sbt test`.
2. Check the JUnit XML output in `target/test-reports`.

## problem

The `<system-out>` and `<system-err>` tags are always empty, even if the unit tests produced output. This is because [`JUnitXmlTestsListener.scala` hardcodes them to always be empty](https://github.com/sbt/sbt/blob/fabeed222799efcc46be15bccf2c31e384f7e369/testing/src/main/scala/sbt/JUnitXmlTestsListener.scala#L166). This makes it harder to track down the cause of unit test failures.

## expectation

Those XML tags should be populated with each unit test's sysout and syserr (or SBT should be configurable so that they can be populated).

## notes

https://github.com/dmitchellmim/sbt-junit-sysout is a simple project to reproduce the issue. It uses Scala 2.13.3 and SBT 1.2.8, but based on the Git history of `JUnitXmlTestsListener.scala`, I think all versions have the same issue.

**Comment by eed3si9n** [comment]
JUnit XML report was added in sbt 0.13.5 as an enhancement to list the test results for CI environment that could interpret it, and I don't think it ever supposed things like capturing standard error or output. From sbt's project's point of view, what we have is ok, and anything more I'd consider it a nice to have. We should also be careful that implementing this will not degrade test performance for existing builds.

**Comment by ashleymercer** [comment]
I'm happy to have a go at fixing this - @eed3si9n would an acceptable solution be to add new flags `testCaptureStdOut` and `testCaptureStdErr` to the configuration? These would be `false` by default to be backward-compatible.

**Comment by eed3si9n** [comment]
> I'm happy to have a go at fixing this - @eed3si9n would an acceptable solution be to add new flags `testCaptureStdOut` and `testCaptureStdErr` to the configuration? These would be `false` by default to be backward-compatible.

That sounds good to me.

**Comment by ashleymercer** [comment]
Okay, I've looked at this a bit more and it's sadly not as simple as I first thought:

- sbt itself can [fork one or more child processes](https://www.scala-sbt.org/1.x/docs/Forking.html#Forking) to actually run tests
- individual test runners might have their own parallelism built in within single test-runner instance: for example, [ScalaTest states](https://www.scalatest.org/user_guide/using_scalatest_with_sbt):

>  ScalaTest uses its own thread pool to run ParallelTestExecution suites, its pool size is determined by the following formula:
> 
>  Pool Size = Available Processors x 2

So we potentially have a situation where a single sbt parent process is sending tests (typically, whole `Suite`s) to runners in multiple JVMs, and those JVMs are each running multiple tests in parallel. It seems that therefore this can't (sadly) be fixed with a simple change to `JUnitXmlTestsListener` like I'd hoped, since the problem is much deeper.

Even in the simple case where everything is on one JVM, and tests run serially, fixing this properly will probably require changes to the individual test frameworks - they would need to be modified to capture standard out and standard error *only* for the duration of a single test (and not anything produced by the test framework itself); and then to somehow feed that back to sbt. (We can't, for example, simply have sbt capture the stdout/err while tests are running, because we have no way of sifting lines and tying them back to individual tests, so there's no guarantee they'd end up in the right report file).

For comparison, it looks like the `junit-interface` plugin for sbt does something like this already, handling cases where it's running in pure-Java or Java+Scala configurations, to buffer standard out/error for a test:

https://github.com/sbt/junit-interface/blob/develop/src/main/java/com/novocode/junit/OutputCapture.java

@eed3si9n thoughts on this? I suspect such a large change would require co-ordination so is probably beyond me on my own.

**Comment by taillred** [comment]
@eed3si9n Can I pick this up?

**Comment by eed3si9n** [comment]
@jakearmstrong59 Sure. Please check out our [Contributor's Guildeline](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md) and:
1. Make sure you can reproduce the problem and the fix manually
2. Include tests for the change
3. Follow our AI assisted contributions guideline
4. Multiple contributors might work on this issue in parallel
5. Generally we recommend getting CI to pass on a forked repo before sending a pull request
6. Sign Scala CLA <https://contribute.akka.io/contribute/cla/scala>
